quic: add TLS session ticket resumption support

bellatoris opened the pull request:
Commit Message: quic: add session ticket resumption support using configured session ticket keys
Additional Description:

## Summary

TLS session resumption is essential for QUIC performance. Without it, every connection requires a full TLS handshake, and 0-RTT becomes meaningless since there's no session state to resume from. As noted in #42682, TLS-related data accounts for roughly 1/3 of bytes during connection establishment - session resumption eliminates most of this overhead.

Currently, Envoy's QUIC implementation does not support session resumption across workers or processes. While users can configure `session_ticket_keys` or `session_ticket_keys_sds_secret_config` in downstream TLS context, these settings have no effect on QUIC connections. This limitation is documented in #25418, which explicitly states that session ticket key plumbing is missing from the QUIC implementation.

This PR bridges that gap by enabling QUIC to use the same session ticket keys configured for TCP TLS, allowing session resumption to work across workers and processes.

## Implementation

We subclass QUICHE's `TlsServerHandshaker` as `EnvoyTlsServerHandshaker` and install a session-ticket key callback on the shared QUICHE `SSL_CTX`. The callback reuses `ServerContextImpl::sessionTicketProcess()` so QUIC and TCP TLS share identical session-ticket handling (same keys, same format, same rotation semantics).

**Key design decisions:**

1. **Per-connection pinning of `ServerContextImpl`**: Each `EnvoyTlsServerHandshaker` holds a `ServerContextSharedPtr` captured at connection creation, and stores `this` in SSL ex_data. The static ticket callback retrieves the handshaker from ex_data and delegates to the pinned context's `sessionTicketProcess()`. Because the shared pointer keeps the context alive, an SDS update that rotates the factory's active context does not invalidate in-flight connections — matching TCP TLS behavior where each connection is bound to the `ServerContextImpl` active at connection creation.

2. **`SSL_CTX_set_tlsext_ticket_key_cb` over `SSL_CTX_set_ticket_aead_method`**: We use the same callback mechanism as TCP TLS rather than QUICHE's `TicketCrypter` interface, so `ServerContextImpl::sessionTicketProcess()` can be reused unchanged.

3. **Graceful fallback**: If the runtime guard is toggled between `OnNewSslCtx` (which installs the callback on the shared `SSL_CTX`) and connection creation (which may fall back to the vanilla handshaker), the ticket callback finds a null handshaker in ex_data and returns 0 to skip ticket issuance for that connection rather than crashing.

## Flow

```
Server startup (once per SSL_CTX):
  EnvoyQuicProofSource::OnNewSslCtx()
    └─ if runtime flag on:
         SSL_CTX_set_tlsext_ticket_key_cb(ssl_ctx, EnvoyTlsServerHandshaker::ticketKeyCallback)

Per connection:
  EnvoyQuicCryptoServerStreamFactoryImpl::createEnvoyQuicCryptoServerStream()
    └─ reads SessionTicketConfig from QuicServerTransportSocketFactory
    └─ constructs EnvoyTlsServerHandshaker(session, crypto_config, factory.sslCtx(), disable_resumption)
         └─ pins ServerContextSharedPtr
         └─ SSL_set_ex_data(ssl, handshakerExDataIndex(), this)
         └─ if disable_resumption || no ticket keys: DisableResumption()  // SSL_OP_NO_TICKET

During handshake (BoringSSL-driven):
  ticketKeyCallback(ssl, ...)
    └─ handshaker = SSL_get_ex_data(ssl, handshakerExDataIndex())
    └─ if null: return 0  // guard toggled after OnNewSslCtx — skip ticket
    └─ return handshaker->pinnedServerContext()->sessionTicketProcess(ssl, ...)
```

Risk Level: Low (behind runtime guard, disabled by default)
Testing: New unit tests for `EnvoyTlsServerHandshaker` and `EnvoyQuicProofSource`; new integration coverage in `sds_dynamic_integration_test` (`SessionTicketKeysViaSds`, `SessionTicketKeysRemovedViaSds`) and in `quic_http_integration_test` (`SessionTicketResumptionWithStaticKeys`, `NoSessionTicketResumptionWithoutKeys`).
Docs Changes: N/A
Release Notes: Added
Platform Specific Features: N/A
[Optional Runtime guard:] `envoy.reloadable_features.quic_session_ticket_support` (default: false)
[Optional Fixes #Issue] Partially addresses #25418

bellatoris committed:
quic: add TLS session ticket resumption support

This change enables TLS session ticket resumption for QUIC connections,
allowing clients to resume TLS sessions without full handshakes when
reconnecting across server instances.

Implementation details:
- Add EnvoyTlsServerHandshaker that provides custom ProofSourceHandle
  to intercept certificate selection and configure session ticket options
- Store filter chain pointer in SSL ex_data during handshake to enable
  session ticket callback to access the correct transport socket factory
- Delegate session ticket encryption/decryption to ServerContextImpl
  which uses configured session_ticket_keys
- Add runtime guard envoy.reloadable_features.quic_session_ticket_support
  (default false) to control feature enablement

The feature integrates with existing DownstreamTlsContext configuration:
- Uses session_ticket_keys from TLS context for ticket encryption
- Respects disable_stateless_session_resumption setting
- Honors handles_session_resumption capability flag

Risk Level: Low (behind runtime guard, disabled by default)

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
fix: format issues

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

kyessenov commented:
Please merge main.
/wait

bellatoris committed:
Resolve merge conflict in changelogs/current.yaml

bellatoris committed:
merge main

bellatoris committed:
merge main

danzh2010 reviewed (commented):
@RyanTheOptimist do you mind take a look at this. I'm not an expert of the TLS handshake implementation.

RyanTheOptimist commented on source/common/quic/quic_server_transport_socket_factory.h:
```diff
@@ -38,6 +38,20 @@ class QuicServerTransportSocketFactory : public Network::DownstreamTransportSock
+  struct SessionTicketConfig {
+    bool has_keys;
+    bool disable_stateless_resumption;
+    bool handles_session_resumption;
```
Please comment on the meaning of these fields. Particular since the latter two both refer to "resumption" it would be good to make the distinction clear.

bellatoris committed:
merge upstream/main

Resolve conflicts:
- changelogs/current.yaml: keep both PR entry and upstream entries
- source/common/quic/BUILD: keep macros dep, remove cert_compression_lib
  (upstream removed), update quiche dep names
- source/common/quic/envoy_quic_proof_source.cc: use registerCertCompression
  (upstream rename), remove cert_compression.h include

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
Address review comments and fix QUICHE API compatibility

- Replace dynamic_cast with static_cast for QuicServerTransportSocketFactory
- Use handshaker->disableTicketSupport() instead of direct SSL_set_options
- Add documentation comments to SessionTicketConfig fields
- Remove unnecessary #include "openssl/ssl.h" from envoy_quic_proof_source.h
- Fix SelectCertificate signature to match upstream QUICHE API change
  (added bool disable_alps_explicit_codepoint parameter)

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
Merge remote-tracking branch 'upstream/main' into doogie/quic-proof-source-update

# Conflicts:
#	changelogs/current.yaml

bellatoris committed:
fix: add 'codepoint' to spelling dictionary

The QUICHE API parameter `disable_alps_explicit_codepoint` triggers
the spelling checker. The plural form `codepoints` already exists.

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
test: add coverage tests for QUIC session ticket resumption

Add tests to improve coverage for source/common/quic and
source/extensions/quic/crypto_stream:

quic_transport_socket_factory_test.cc:
- CreateDownstreamTransportSocketPanics: verify PANIC stub
- GetSessionTicketConfig: verify default config values
- SessionTicketProcessNullCtx: verify null ssl_ctx returns 0

envoy_quic_proof_source_test.cc:
- FilterChainExDataIndex: verify stable non-negative index
- OnNewSslCtxWithSessionTicketSupport: smoke test with guard enabled
- OnNewSslCtxWithSessionTicketSupportDisabled: smoke test with guard disabled

envoy_quic_dispatcher_test.cc:
- HandshakeWithSessionTicketSupport: process CHLO through
  EnvoyTlsServerHandshaker with runtime guard enabled to exercise
  SelectCertificate, ComputeSignature, and disableTicketSupport

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

RyanTheOptimist commented on source/common/quic/envoy_tls_server_handshaker.cc:
```diff
@@ -0,0 +1,171 @@
+    auto ticket_config = factory.getSessionTicketConfig();
+    if (ticket_config.disable_stateless_resumption || !ticket_config.has_keys ||
+        ticket_config.handles_session_resumption) {
+      SSL_set_options(client_hello.ssl, SSL_OP_NO_TICKET);
```
Ok, once https://github.com/envoyproxy/envoy/pull/44019 lands, a non-const `tls_connection()` method will be available.

/wait

bellatoris committed:
test: add coverage tests for QUIC session ticket resumption

Add tests to improve coverage for source/common/quic and
source/extensions/quic/crypto_stream:

Source changes:
- Remove custom disableTicketSupport() wrapper, use QUICHE's new
  non-const tls_connection().DisableTicketSupport() API instead
- Remove unnecessary openssl/ssl.h include from handshaker header

quic_transport_socket_factory_test.cc:
- CreateDownstreamTransportSocketPanics: verify PANIC stub
- GetSessionTicketConfig: verify default config values
- SessionTicketProcessNullCtx: verify null ssl_ctx returns 0

envoy_quic_proof_source_test.cc:
- FilterChainExDataIndex: verify stable non-negative index
- OnNewSslCtxWithSessionTicketSupport: smoke test with guard enabled
- OnNewSslCtxWithSessionTicketSupportDisabled: smoke test with guard disabled

envoy_quic_dispatcher_test.cc:
- HandshakeWithSessionTicketSupport: process CHLO through
  EnvoyTlsServerHandshaker with runtime guard enabled to exercise
  SelectCertificate and session ticket configuration paths

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
test: extract ticketKeyCallback for testability and add callback coverage tests

Extract the stateless lambda in OnNewSslCtx to a public static method
ticketKeyCallback so it can be unit-tested directly. Use dynamic_cast
(consistent with getTransportSocketAndFilterChain in the same file).

Add two new tests:
- TicketKeyCallbackNullFilterChain: verifies return 0 when no filter chain
  ex_data is set on the SSL
- TicketKeyCallbackWithFilterChain: verifies delegation through to the
  transport socket factory sessionTicketProcess

Also extract loadCertsIntoFactory() helper to eliminate 3x copy-pasted
cert-loading mock setup boilerplate across tests.

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
test: add old GetCertChain API path coverage for handshaker SelectCertificate

Add HandshakeWithOldCertChainAPI test that disables the QUICHE
quic_use_proof_source_get_cert_chains flag (default true) to exercise
the old SelectCertificate code path using GetCertChain
(envoy_tls_server_handshaker.cc L124-138). The existing test covers
the new GetCertChains path (L104-121) with the flag at default.

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

RyanTheOptimist commented on source/common/quic/envoy_tls_server_handshaker.h:
```diff
@@ -0,0 +1,65 @@
+  ~EnvoyTlsServerHandshaker() override = default;
+
+protected:
+  std::unique_ptr<quic::ProofSourceHandle> MaybeCreateProofSourceHandle() override;
```
Thanks for refactoring to use TlsConnection. This is much clearer now. If I understand things correctly, the only override we have is this one to create the `ProofSourceHandle`. Envoy already has a factory for creating proof sources. Is that something we could take advantage of in this case?

https://github.com/envoyproxy/envoy/blob/main/source/common/quic/envoy_quic_proof_source_factory_interface.h

bellatoris committed:
refactor: eliminate custom EnvoyTlsServerHandshaker, use QUICHE default

Remove EnvoyTlsServerHandshaker and its custom EnvoyProofSourceHandle.
QUICHE's DefaultProofSourceHandle already has identical SelectCertificate
and ComputeSignature logic. The only Envoy-specific behavior — setting
filter chain in SSL ex_data and conditionally disabling session tickets —
is moved to createEnvoyQuicCryptoServerStream using QUICHE's public API:
- GetSsl() for SSL_set_ex_data
- DisableResumption() for ticket control (works at construction time
  before EarlySelectCertCallback sets can_disable_resumption_ to false)

The filter chain pointer from position_->filter_chain_ is safe for the
session's lifetime (closeConnectionsWithFilterChain destroys sessions
synchronously before freeing filter chains).

Also rename sessionTicketProcess -> processSessionTicket per reviewer nit.

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

Merge upstream main

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
Merge upstream main

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris commented on source/common/quic/envoy_tls_server_handshaker.cc:
```diff
@@ -0,0 +1,171 @@
+        ticket_config.handles_session_resumption) {
+      handshaker_->tls_connection().DisableTicketSupport();
+    }
+  }
```
https://github.com/envoyproxy/envoy/pull/42734#discussion_r2978577984, thank you for suggestion.

bellatoris committed:
fix: spelling check error in dispatcher test comment

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

RyanTheOptimist reviewed (commented):
Thanks for the great suggestion @danzh2010. This looks great, @bellatoris! But let's wait for @danzh2010's LGTM before landing.

bellatoris committed:
address review: use transport_socket_factory directly, remove filter_chain param

Per danzh2010's review:
- Use existing transport_socket_factory param instead of adding filter_chain
- Store transport_socket_factory ptr in SSL ex_data instead of filter_chain
- Remove early return in OnNewSslCtx, wrap single line in if block
- Use static_pointer_cast for ServerContextImpl in processSessionTicket

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
Merge upstream main

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

RyanTheOptimist reviewed (dismissed):
Thanks for all the work on the PR. It's looking great, and super clear!

bellatoris committed:
address review: ENVOY_BUG for null factory, rename test, replace semi-integration test with e2e

Per danzh2010's review:
- Use IS_ENVOY_BUG when factory is null in ticketKeyCallback
- Rename FilterChainExDataIndex test to TransportSocketFactoryExDataIndex
- Remove EnvoyQuicDispatcherWithSessionTicketTest semi-integration test
- Add e2e integration tests in quic_http_integration_test.cc for session
  ticket resumption with static keys and without keys

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
fix: guard against empty session ticket keys in processSessionTicket

When session ticket keys are removed via SDS update while a QUIC connection
is active, the factory's ssl_ctx_ is swapped to a new ServerContextImpl with
empty keys. Unlike TCP TLS where each connection pins its own context, QUIC
reads the current ssl_ctx_ which may have empty keys, hitting the
RELEASE_ASSERT in sessionTicketProcess. Guard with hasSessionTicketKeys()
on the same object to prevent the crash and fall back to a full handshake.

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
fix: use inline_bytes for session ticket key in integration test

The CI remote execution environment cannot resolve file paths via
TestEnvironment::runfilesPath. Use inline_bytes with a raw 80-byte
key instead.

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
test: add SDS session ticket integration tests

Add two e2e integration tests for session ticket keys delivered via SDS:
- SessionTicketKeysViaSds: verifies connections succeed when keys are
  delivered via SDS
- SessionTicketKeysRemovedViaSds: verifies no crash when keys are removed
  via SDS update (the scenario mayfieldiv identified)

Both tests run for TCP and QUIC via existing parameterization.

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris commented:
/retest

danzh2010 commented on test/integration/sds_dynamic_integration_test.cc:
```diff
@@ -233,6 +234,39 @@ class SdsDynamicDownstreamIntegrationTest : public SdsDynamicIntegrationBaseTest
+          sds_config->set_name(session_ticket_keys_secret_);
+          auto* config_source = sds_config->mutable_sds_config();
+          config_source->mutable_path_config_source()->set_path(session_ticket_keys_sds_path_);
+          config_source->set_resource_api_version(envoy::config::core::v3::ApiVersion::V3);
```
nits: move this shared code into a helper function?

danzh2010 commented on test/integration/sds_dynamic_integration_test.cc:
```diff
@@ -615,6 +685,49 @@ TEST_P(SdsDynamicDownstreamIntegrationTest, WrongSecretFirst) {
+      listenerStatPrefix(test_quic_
+                             ? "quic_server_transport_socket_factory.context_config_update_by_sds"
+                             : "server_ssl_socket_factory.ssl_context_update_by_sds"),
+      2);
```
Let's not waiting for the sds update and let the 2nd request race with SDS. Please run the test 100 times to make sure it doesn't crash.

danzh2010 commented on source/common/quic/envoy_quic_proof_source.h:
```diff
@@ -27,6 +27,26 @@ class EnvoyQuicProofSource : public EnvoyQuicProofSourceBase {
+  static int ticketKeyCallback(SSL* ssl, uint8_t* key_name, uint8_t* iv, EVP_CIPHER_CTX* ctx,
+                               HMAC_CTX* hmac_ctx, int encrypt);
+
+  struct TransportSocketFactoryWithFilterChain {
+    const QuicServerTransportSocketFactory& transport_socket_factory_;
+    const Network::FilterChain& filter_chain_;
+  };
+
+  absl::optional<TransportSocketFactoryWithFilterChain>
+  getTransportSocketAndFilterChain(const quic::QuicSocketAddress& server_address,
+                                   const quic::QuicSocketAddress& client_address,
+                                   const std::string& hostname);
+
```
Do these still need to be public?

danzh2010 replied:
Is this addressed?

bellatoris replied:
yup

bellatoris replied:
Restored the original member order in the private section — this file now has zero diff against main. @danzh2010 PTAL

danzh2010 commented on source/common/quic/quic_server_transport_socket_factory.cc:
```diff
@@ -175,6 +175,24 @@ QuicServerTransportSocketFactory::getTlsCertificateAndKey(absl::string_view sni,
+  // QuicServerTransportSocketFactory always creates ServerContextImpl.
+  auto server_ctx =
+      std::static_pointer_cast<Extensions::TransportSockets::Tls::ServerContextImpl>(ssl_ctx);
+  return server_ctx->sessionTicketProcess(ssl, key_name, iv, ctx, hmac_ctx, encrypt);
```
Nice cache! Thanks for thinking through this corner case! Is it possible to align with SSL socket's lifetime by retaining the `ServerContextImpl` object somewhere?  Instead of storing transport socket factory in SSL object via `SSL_set_app_data()`, create a custom handshaker which retains a shared pointer to`ServerContextImpl` and retaining handshaker via `SSL_set_app_data()`.

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
refactor: consolidate session ticket logic into EnvoyTlsServerHandshaker

Address PR review comments by creating a custom TlsServerHandshaker subclass
that pins ServerContextImpl at connection creation time. This consolidates
session ticket logic that was spread across EnvoyQuicProofSource,
QuicServerTransportSocketFactory, and the crypto stream factory into a
single self-contained class.

Key changes:
- New EnvoyTlsServerHandshaker: owns pinned context, ex_data index, callback
- Remove processSessionTicket from QuicServerTransportSocketFactory
- Remove transportSocketFactoryExDataIndex and ticketKeyCallback from ProofSource
- Crypto stream no longer depends on ProofSource
- Extract SDS config helper in integration test
- Remove waitForCounterGe in SDS race test (let 2nd request race)

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

RyanTheOptimist commented:
/wait

bellatoris committed:
fix: spelling check - remove tlsext from comment

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
fix: spelling check - avoid BoringSSL type names in comments

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris commented:
/retest

danzh2010 commented on source/common/quic/envoy_tls_server_handshaker.h:
```diff
@@ -0,0 +1,47 @@
+namespace Envoy {
+namespace Quic {
+
+// Minimal TlsServerHandshaker subclass that pins a ServerContextImpl at
```
Can you clarify what "pin" means here? Especially mention about the life time of ServerContext and why it needs to be cached here.

bellatoris replied:
Expanded the class comment to explain what pinning means, the ServerContext lifetime, and why the pointer needs to be held for the connection's lifetime. @danzh2010 PTAL

danzh2010 commented on test/common/quic/quic_transport_socket_factory_test.cc:
```diff
@@ -113,6 +113,55 @@ TEST_F(QuicServerTransportSocketFactoryConfigTest, ClientAuthUnsupported) {
                             "TLS Client Authentication is not supported over QUIC");
 }
 
+TEST_F(QuicServerTransportSocketFactoryConfigTest, CreateDownstreamTransportSocketPanics) {
```
This is irrelevant to this PR but good to have. Thanks for adding such test! Can you comment that QUIC transport socket factory is not supposed to be used to create transport socket?

bellatoris replied:
Added a comment above the test explaining that QuicServerTransportSocketFactory implements DownstreamTransportSocketFactory only to be stored on a FilterChain — not to create transport sockets, since QUIC connections use the QUICHE stack directly via EnvoyQuicServerSession. @danzh2010 PTAL

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

danzh2010 commented on test/integration/sds_dynamic_integration_test.cc:
Thanks for adding test coverage for SDS with session ticket.

danzh2010 commented on test/common/quic/envoy_quic_proof_source_test.cc:
```diff
@@ -232,6 +212,34 @@ class EnvoyQuicProofSourceTest : public ::testing::Test {
     ASSERT_TRUE(callback_status.ok());
   }
 
+  void expectCertChainAndPrivateKey(const std::string& cert, bool expect_private_key,
```
nits: moving this up for cleaner diff

bellatoris replied:
Moved expectCertChainAndPrivateKey back above loadCertsIntoFactory for a cleaner diff. @danzh2010 PTAL

bellatoris committed:
fix: address review comments (comment clarifications + test cleanup)

- Expand EnvoyTlsServerHandshaker class comment to explain pinning rationale,
  ServerContext lifetime, and how it matches TCP TLS behavior
- Restore original order of private members in envoy_quic_proof_source.h
  (zero diff against main for this file)
- Add comment on CreateDownstreamTransportSocketPanics explaining why the
  QUIC transport socket factory is not used to create transport sockets
- Swap order of expectCertChainAndPrivateKey and loadCertsIntoFactory for
  cleaner diff against main

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
fix: spelling check - avoid QUICHE's possessive in comment

The CI spelling checker accepts QUICHE (in tools/spelling/spelling_dictionary.txt)
but not the possessive form QUICHE's. Rewrote the sentence to avoid it.

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

RyanTheOptimist commented:
Are the comments from @danzh2010 addressed?
Looks like this needs a main merge.
/wait

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris commented:
> Are the comments from @danzh2010 addressed?

yes

> Looks like this needs a main merge.

done

bellatoris commented:
/retest

danzh2010 reviewed (dismissed):
LGTM, thanks for contributing to QUIC stack!

danzh2010 commented:
Please update the PR description with new implementation approach

bellatoris commented:
> Please update the PR description with new implementation approach

@danzh2010 Done, thank you for the review.

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris commented:
Looks like current.yaml was merged incorrectly in this PR. https://github.com/envoyproxy/envoy/pull/44287

bellatoris commented:
/retest

danzh2010 commented:
@RyanTheOptimist mind doing another pass? The code has diverged quite a bit since your last stamp.

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris commented:
/retest

RyanTheOptimist reviewed (dismissed):
LGTM

RyanTheOptimist commented:
Looks like this needs a main merge
/wait

bellatoris committed:
Merge branch 'main' into doogie/quic-proof-source-update

Signed-off-by: Doogie Min <doogie.min@sendbird.com>

bellatoris commented:
> Looks like this needs a main merge /wait

Done
